src/snowflake/snowpark/mock/_plan.py

```diff
@@ -1379,6 +1389,98 @@ def outer_join(base_df):
+            )
+
+        pivot_column = plan.session._analyzer.analyze(source_plan.pivot_column)
+
+        if isinstance(source_plan.pivot_values, Iterable):
+            pivot_values = [exp.value for exp in source_plan.pivot_values]
+        elif source_plan.pivot_values is None:
+            pivot_values = []
+        else:
+            analyzer.session._conn.log_not_supported_error(
+                external_feature_name=f"Pivot values from {source_plan.pivot_values}",
+                internal_feature_name=type(source_plan).__name__,
+                raise_error=NotImplementedError,
+            )
+
+        # source_plan.grouping_columns contains columns specified in the groupby clause.
+        # If that clause is omitted then the default behavior is to use all columns that are not
+        # being used as either the agg column or the pivot column.
+        # See here for more details: https://community.snowflake.com/s/article/Pivot-returns-more-rows-than-expected
+        grouping_columns = [
+            plan.session._analyzer.analyze(c) for c in source_plan.grouping_columns
+        ]
+        indices = grouping_columns or [
+            col for col in child_rf.keys() if col not in {agg_column, pivot_column}
+        ]
+
+        # Missing values are filled with a sentinel object that can later be replaced with Nones
+        sentinel = object()
+
+        # Snowflake treats an empty aggregation as None, whereas pandas treats it as 0.
+        # This requires us to wrap the aggregation function with extract logic to handle this special case.
+        def agg_function(column):
+            return (
+                agg_functions[agg_function_name](column.dropna())
+                if column.any()
+                else sentinel
+            )
+
+        default = (
+            source_plan.default_on_null.value if source_plan.default_on_null else None
+        )
+
+        # Count defaults to 0 rather than None
+        if agg_function_name == "count":
+            default = default or 0
+
+        result = child_rf.pivot_table(
+            columns=pivot_column,
+            values=agg_column,
+            aggfunc=agg_function,
+            index=indices,
+        )
+        result.reset_index(inplace=True)
+
+        # Select down to indices and provided values if specific values were requested
+        if pivot_values:
+            result = result[list(indices) + pivot_values]
+
+        # Non-indice columns lack an sf_type, add them back in.
+        for res_col in set(result.columns) - set(indices):
+            # fill_na will not fill na like values with None, but it can replace them with a sentinel value
+            filled = result[res_col].fillna(sentinel)
+
+            # Sentinel values are replaced with None, then all Nones are replaced with the default if provided
+            data = filled.replace({sentinel: None}).replace({None: default}).values
```

Comment:
Is this different than `result[res_col].fillna(default)`?

Reply:
There can be Nones in the dataframe if there were Nones in the source data. There can also be sentinel values generated by the aggregation functions. I could probably refactor to go directly to Nones skipping the sentinel value altogether. 

Originally I wanted to use the sentinel value as a fill value for the pivot_table call so that I could preserve values that were NaN for reasons unrelated to the pivot, but a recent change in pandas prevents both using None as a fill value and using the object sentinel value trick. This caused a test failure in my original approach since I was on pandas 2.0.0 and the test env was on 2.2.2